### PR TITLE
internal/store/pod.go: Only create waiting_reason series if pods are in waiting state

### PR DESCRIPTION
**Pull request by lilic**
Due to the extremely high cardinality of this metric and with 2.0 release being breaking, this now changes that the waiting_reason metrics only include series that have state `Waiting`. We also now can include all the reasons as cardinality will only always be O(count of containers waiting). I made the adjustment for both containers and init containers.

The reduction of series can be seen in the unit tests, but I also tried it on my cluster and this is the result we only see the waiting containers:

```
# HELP kube_pod_container_status_waiting_reason Describes the reason the container is currently in waiting state.
# TYPE kube_pod_container_status_waiting_reason gauge
kube_pod_container_status_waiting_reason{namespace="default",pod="nginx-deployment-6b99d49f64-km96g",container="nginx",reason="ImagePullBackOff"} 1
kube_pod_container_status_waiting_reason{namespace="default",pod="nginx-deployment-6b99d49f64-2ghjb",container="nginx",reason="ImagePullBackOff"} 1
kube_pod_container_status_waiting_reason{namespace="default",pod="nginx-deployment-6b99d49f64-pgj2t",container="nginx",reason="ImagePullBackOff"} 1
```
Will open PRs to adjust the 0 values for other metrics here as well in another PR.

@brancz @smarterclayton @vsxen can you please have a look if this metric continues to work as you want.

Fixes https://github.com/kubernetes/kube-state-metrics/issues/1321

### CODE REVIEW

**Comment by brancz** [comment]
Seems like CI is not happy, but strategy and implementation is sound.

/lgtm

**Comment by lilic** [comment]
Thanks @brancz I always forget about our main tests :D PTAL, thanks!

/hold

Put a holt until one of the reports belives this fixes it 👍

File: docs/pod-metrics.md [hunk 1]
```diff
@@ -12,7 +12,7 @@
...
 | kube_pod_container_info | Gauge | `container`=&lt;container-name&gt; <br> `pod`=&lt;pod-name&gt; <br> `namespace`=&lt;pod-namespace&gt; <br> `image`=&lt;image-name&gt; <br> `image_id`=&lt;image-id&gt; <br> `container_id`=&lt;containerid&gt; | STABLE |
 | kube_pod_container_status_waiting | Gauge | `container`=&lt;container-name&gt; <br> `pod`=&lt;pod-name&gt; <br> `namespace`=&lt;pod-namespace&gt; | STABLE |
-| kube_pod_container_status_waiting_reason | Gauge | `container`=&lt;container-name&gt; <br> `pod`=&lt;pod-name&gt; <br> `namespace`=&lt;pod-namespace&gt; <br> `reason`=&lt;ContainerCreating\|CrashLoopBackOff\|ErrImagePull\|ImagePullBackOff\|CreateContainerConfigError\|InvalidImageName\|CreateContainerError&gt; | STABLE |
+| kube_pod_container_status_waiting_reason | Gauge | `container`=&lt;container-name&gt; <br> `pod`=&lt;pod-name&gt; <br> `namespace`=&lt;pod-namespace&gt; <br> `reason`=&lt;ContainerCreating\|CrashLoopBackOff\|ErrImagePull\|ImagePullBackOff\|CreateContainerConfigError\|InvalidImageName\|CreateContainerError&gt; | EXPERIMENTAL |
```
Comment (lilic) [inline docs/pod-metrics.md:15]: As this willd include but is not anymore limited to just these reasons, not sure if removing them makes sense or not 🤔  @brancz any suggestions?

Comment (brancz) [reply docs/pod-metrics.md:15]: Yeah let's just put a generic placeholder here.

**Comment by smarterclayton** [comment]
Looks great, thanks!

**Comment by lilic** [comment]
/hold cancel

@brancz PTAL, I adjusted the docs so lgtm got dropped.

**Comment by vsxen** [comment]
LGTM

**Pull request merged**